### DISCUSSION TITLE
How are the specific versions of the tools in the toolchain included in the full image selected?

### DISCUSSION THREAD

**Discussion by jackson-chris** [renovatebot/renovate#28357] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Github, renovate version 37.177.0-full

### Please tell us more about your question or problem

We have a newer repo were targeting with renovatebot and it uses gradle but has a requirement to use java 21, the `build.gradle.kts` file has the following:
```
java {
    toolchain {
        languageVersion.set(JavaLanguageVersion.of(21))
        vendor.set(JvmVendorSpec.AZUL)
    }
}
```

and renovate PR reports a failure:

```
* What went wrong:
Could not resolve all dependencies for configuration ':agent'.
> Failed to calculate the value of task ':compileJava' property 'javaCompiler'.
   > Cannot find a Java installation on your machine matching this tasks requirements: {languageVersion=21, vendor=AZUL, implementation=vendor-specific} for LINUX on x86_64.
      > No locally installed toolchains match and toolchain download repositories have not been configured.
```

A few questions about this:

1. The renovatebot full image includes toolschains OOTB, but what is currently dictating the version. I see https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/maven-wrapper/artifacts.spec.ts#L213 which is what I think explains the version I see in the container image:
    ```
    ± docker run -ti --rm renovate/renovate:37.177.0-full /bin/bash
    ubuntu@4a22a6b19652:/usr/src/app$ java --version
    openjdk 17.0.10 2024-01-16
    OpenJDK Runtime Environment Temurin-17.0.10+7 (build 17.0.10+7)
    OpenJDK 64-Bit Server VM Temurin-17.0.10+7 (build 17.0.10+7, mixed mode, sharing)
    ```
    is that correct? or something else I'm missing?
2. If java 21 is NOT currently included what is the appropriate way to request it's added to the image?
3. Is adding the specific java version to the renovate image the typical solution or do you suggest we try and manage that via [gradle functionality](https://docs.gradle.org/8.7/userguide/toolchains.html#sec:provisioning).  If this is the recommended approach it's easy to configure however we are concerned about having to repeatedly download the JDK for each and every renovatebot job we run as the majority of our repos are starting to swap over to java 21. There is also the concern of rate limiting or being added to a disallow list if we request it too often. 

NOTE: I did check the latest version but still looks like it's using java 17.

### Logs (if relevant)

See above.

**Comment by rarkins** [comment]
Generally speaking we'll install the latest stable or LTS version of tools into the full image, however we won't do major upgrades unless it's a major upgrade of Renovate. The upcoming v38 of Renovate will include such major updates, although @viceice it looks like we missed Java in https://github.com/renovatebot/base-image/releases/tag/3.0.0

**Reply by rarkins** [reply]
Here is a fix for why v21 wasn't being suggested to us: https://github.com/renovatebot/renovate/pull/28361
